engine/include/tsengine/math.hpp

```diff
@@ -71,15 +71,191 @@ inline Vec4::Vec4(const float x_, const float y_, const float z_, const float w_
     x{x_}, y{y_}, z{z_}, w{w_}
 {}
 
-struct BaseMatrix {};
-
 template<size_t matRows, size_t matColumns>
-struct Matrix : BaseMatrix
+struct Matrix
 {
     static constexpr size_t rowsNum{matRows};
     static constexpr size_t colsNum{matColumns};
 };
 
+struct Mat2 : public Matrix<2, 2>
+{
+    std::array<Vec2, 2> data{};
+
+    Mat2()
+    {}
+
+    Mat2(
+        const float x1, const float y1,
+        const float x2, const float y2)
+        :
+        data
+        {{
+            {x1, y1},
+            {x2, y2}
+        }}
+    {}
+
+    explicit Mat2(const float value) :
+        data
+    {{
+        {value, 0.f  },
+        {0.f  , value}
+    }}
+    {}
+
+    Vec2& operator[](const size_t index)
+    {
+        if ((index < 0) || (index > 2))
+        {
+            throw std::runtime_error("invalid index specified");
+        }
+
+        return data[index];
+    }
+
+    const Vec2& operator[](const size_t index) const
+    {
+        if ((index < 0) || (index > 2))
+        {
+            throw std::runtime_error("invalid index specified");
+        }
+
+        return data[index];
+    }
+};
+
+inline Mat2 operator*(const Mat2& lhs, const Mat2& rhs)
+{
+    return
+    {
+        rhs.data[0].x * lhs.data[0].x + rhs.data[0].y * lhs.data[1].x,
+        rhs.data[0].x * lhs.data[0].y + rhs.data[0].y * lhs.data[1].y,
+
+        rhs.data[1].x * lhs.data[0].x + rhs.data[1].y * lhs.data[1].x,
+        rhs.data[1].x * lhs.data[0].y + rhs.data[1].y * lhs.data[1].y,
+    };
+}
+
+inline Vec2 normalize(const Vec2 vec)
+{
+    const auto mag = std::sqrt(vec.x * vec.x + vec.y * vec.y);
+    return {vec.x / mag, vec.y / mag};
+}
+
+inline std::string to_string(const Mat2 mat)
+{
+    return std::format("\n"
+        "{:>10.7f}, {:>10.7f},\n"
+        "{:>10.7f}, {:>10.7f}\n",
+        mat[0].x, mat[0].y,
+        mat[1].x, mat[1].y);
+}
+
+struct Mat3 : public Matrix<3, 3>
+{
+    std::array<Vec3, 3> data{};
+
+    Mat3()
+    {}
+
+    Mat3(
+        const float x1, const float y1, const float z1,
+        const float x2, const float y2, const float z2,
+        const float x3, const float y3, const float z3)
+        :
+        data
+        {{
+            {x1, y1, z1},
+            {x2, y2, z2},
+            {x3, y3, z3}
+        }}
+    {}
+
+    Mat3(const Quat& quat) :
+        data
+        {{
+            {1 - 2 * (quat.y * quat.y + quat.z * quat.z), 2 * (quat.x * quat.y + quat.z * quat.w)    , 2 * (quat.x * quat.z - quat.y * quat.w)    },
+            {2 * (quat.x * quat.y - quat.z * quat.w)    , 1 - 2 * (quat.x * quat.x + quat.z * quat.z), 2 * (quat.y * quat.z + quat.x * quat.w)    },
+            {2 * (quat.x * quat.z + quat.y * quat.w)    , 2 * (quat.y * quat.z - quat.x * quat.w)    , 1 - 2 * (quat.x * quat.x + quat.y * quat.y)},
+        }}
+    {}
+
+    explicit Mat3(const float value) :
```

Comment:
Is it a kind of industry standard in 3D gaming that initializing a matrix with single value gives you identity matrix multiplied by this value? It seems a bit counterintuitive to me.

Reply:
Kinda, we imitate glm library

Reply:
> Is it a kind of industry standard in 3D gaming that initializing a matrix with single value gives you identity matrix multiplied by this value? It seems a bit counterintuitive to me.

83 page
https://registry.khronos.org/OpenGL/specs/gl/GLSLangSpec.4.20.pdf

Reply:
Because on the other hand glm imitates glsl